src/interface/mem.rs

```diff
@@ -0,0 +1,170 @@
+use libc::{MAP_ANONYMOUS, MAP_FIXED, MAP_PRIVATE, PROT_EXEC};
+
+use crate::safeposix::cage::{MemoryBackingType, Vmmap, VmmapOps, MAP_SHARED, PAGESHIFT, PAGESIZE, PROT_NONE, PROT_READ, PROT_WRITE};
+
+use crate::interface::{cagetable_getref, syscall_error, Errno};
+
+pub fn round_up_page(length: u64) -> u64 {
+    if length % PAGESIZE as u64 == 0 {
+        length
+    } else {
+        ((length / PAGESIZE as u64) + 1) * PAGESIZE as u64
+    }
+}
+
+pub fn fork_vmmap(parent_vmmap: &Vmmap, child_vmmap: &Vmmap) {
+    let parent_base = parent_vmmap.base_address.unwrap();
+    let child_base = child_vmmap.base_address.unwrap();
+
+    // iterate through each vmmap entry
+    for (_interval, entry) in parent_vmmap.entries.iter() {
+        // if the entry has PROT_NONE, that means the entry is currently not used
+        if entry.prot == PROT_NONE { continue; }
+        // translate page number to user address
+        let addr_st = (entry.page_num << PAGESHIFT) as i32;
+        let addr_len = (entry.npages << PAGESHIFT) as usize;
+
+        // translate user address to system address
+        let parent_st = parent_vmmap.user_to_sys(addr_st);
+        let child_st = child_vmmap.user_to_sys(addr_st);
+        if entry.flags & (MAP_SHARED as i32) > 0 {
+            // for shared memory, we are using mremap to fork shared memory
```

Comment:
I think in our nacl code theres a longish comment about why we do this, can we move that to here.